### ISSUE TITLE
can't generate a declaration for only a single file when it has dependencies

### ISSUE THREAD

**Issue by dgoldstein0** [Microsoft/TypeScript#13292]
**TypeScript Version:**  2.1.4
**Code**

./modules/animals.ts
```ts
import * as dog from "modules/dog";

export let animal = "gorilla";
export let food = "banana";

export let otheranimal = dog.animal;
```

./modules/dog.ts
```ts
interface dogsound {
  name: "bark" | "woof";
}

export let animal = "dog";
export let food = "dogfood";
export let noise: dogsound = {name: "bark"}

```

./tsconfig.json
```json

{
  "compilerOptions": {
    "module": "amd",
    "moduleResolution": "classic",
    "target": "es5",
    "baseUrl": ".",
    "paths": {"*": ["*"]},
    "outDir": "../compiled"
  },
  "compileOnSave": false,
  "include": [
    "modules/animals.ts"
  ]
}
```

(This compiles with `tsc --project ./tsconfig.json`)

**Expected behavior:**

There should be some way to generate a declaration for just `modules/animals.ts`.  

**Actual behavior:**

Running with `tsc --project tsconfig.json --declaration` almost works (needs dogsound to be exported) but (a) can generate errors that don't exist without --declaration and (b) generates declarations for both typescript files; running with `--noResolve` also, in an attempt to avoid compiling modules/dog.ts, gives the error "cannot find modules/dog".  Which is annoying because if my goal is to just generate a declaration for `modules/animals.ts`, none of the types in `modules/dog` are actually needed for `modules/animals` exports.

While I could just accept that I can generate declarations from all files, generating declarations means several more rules which aren't normally applied by tsc come into force - types which are part of the external interface of a module are required to be exported (TS4025, TS4028, TS4033, TS4055, TS4060, TS4063, TS4073 TS4075, TS4078), which otherwise is not necessary.  If fixing these were trivial, maybe then I wouldn't mind, but on my medium size codebase (~400 typescript files), this resulted in >700 compiler errors, and when I scoped down to just the files that were dependencies of the file I wanted to generate the declaration for, I still had about 30 errors I had to fix.

**Comment by mhegazy** [comment]
> Which is annoying because if my goal is to just generate a declaration for modules/animals.ts, none of the types in modules/dog are actually needed for modules/animals exports.

this is not true in the general case. consider:

```ts
import * as dog from "modules/dog";

export function makeNoise() {
    return dog.noise;
}
```

For the compiler to be able to write the declaration, it needs to know the return type of `makeNoise`, and to do that, it needs to know the type of `dog.noise`.

Writing declaration files is not a syntactic transformation that can operate on a "single file"; it is a needs semantic information, and thus, operates on a "program".

> While I could just accept that I can generate declarations from all files, generating declarations means several more rules which aren't normally applied by tsc come into force - types which are part of the external interface of a module are required to be exported

the idea here is generating declarations requires some additional rules. we could enforce them on every compilation, even ones that will never generate a declaration file, or do it on a pay-per-play fashion, where only compilations with --declaration get the errors.

**Comment by dgoldstein0** [comment]
> Writing declaration files is not a syntactic transformation that can
operate on a "single file"; it is a needs semantic information, and thus,
operates on a "program".

Sure, that makes sense.  But in general most of my types do not leak out
into the interfaces of other modules; I would like a way for these errors
to show up lazily, such that if the type isn't needed for the declaration
of the file I care about, I don't get errors about it not being exported.
Perhaps this could be an option?

**Comment by aluanhaddad** [comment]
Why not just use a an object literal type or a type alias. If the interface is meaningful, just export it.

**Comment by dgoldstein0** [comment]
@aluanhaddad maybe I should explain my use case.  I'm trying to take a subset of my typescript code, and bundle it with webpack for consumption in another one of our codebases which also happens to use typescript.  Sharing typescript source is likely to cause trouble since the two codebases have different compilation processes, hence part of the reason for webpack.  But we don't want to lose our type definitions by doing this, so I'm trying to generate a `.d.ts` for what's going to be the main module in the webpack bundle.  There are two open source tools for doing this - dts-bundle and dts-generator - and both rely on running the typescript compiler with `--declaration`.

Exporting a type is useless since we are handing over a compiled artifact.

Anyhow I wanted to start this discussion not because I'm blocked in this particular instance by all the extra errors that `--declaration` raises - I just took an hour or so and fixed those - but rather I think we'll be doing more of this in the future, and it would be nice to make this process lower friction.

**Comment by aluanhaddad** [comment]
> Exporting a type is useless since we are handing over a compiled artifact.

I don't quite follow. Since you plan to distribute the declarations along with the compiled artifact. What I mean by "export a type" is export it from your TypeScript source file. As you know it will be emitted in the `.d.ts` files and not in the JavaScript code.



> There are two open source tools for doing this - dts-bundle and dts-generator - and both rely on running the typescript compiler with --declaration.

Well, I'm not familiar with these tools. I would just include all of the declaration files emitted by a single plain run of `tsc --declaration`. Does your Webpack bundle have a different API surface than the individual files that comprise it? If not, if bundling is simply a distribution/implementation detail, then you don't need to generate a `.d.ts` _bundle_ and can just distribute those files as part of your package as they are laid out prior to bundling. 
A lot of libraries do this. 
Also, note that it is worth distributing the source files (.js) of your bundle as individual artifacts so that users with different tool chains can consume them as they see fit. 
None of this conflicts (in theory) with distributing a bundle.

**Comment by dgoldstein0** [comment]
> As you know it will be emitted in the .d.ts files and not in the
JavaScript code.

Actually didn't realize that but I guess it's true.  Anyhow I suppose what
I'm asking for is a way to generate only the parts relevant to a particular
declaration file; right now you end up enforcing that all necessary types
are exported from every module, which isn't ideal when some of them aren't
exposed from the one module that makes up the final interface to the bundle.

**Comment by aluanhaddad** [comment]
So you're saying that you're bundle doesn't support deep imports? I think bundling is a bad principle around which to scope and design an API surface. What if someone wanted to use your package in nodejs?

**Comment by dgoldstein0** [comment]
This particular code I'm working on isn't going to be public in npm, and
definitely doesn't need to support deep imports.  I expect it'll have a
single digit number of uses.

Given that I don't need deep imports, does what I'm asking for make sense?

**Comment by aluanhaddad** [comment]
I see what you're saying but since it's an internal package it matters even less, in my opinion, that you constrain the API surface. What I'm saying is that I would avoid specifying your interfaces based on how your code is packaged by a specific bundler.

**Comment by dgoldstein0** [comment]
let me rephrase the problem: a large part of what we are trying to do here is make it possible to build different projects with different build tools - ultimately using typescript under the hood - and make one project depend on another, but without (1) requiring the reused code to be compatible (i.e. compiled from scratch) with multiple build tools, and (2) without losing our types.

Anyhow I think it would be useful for typescript to be able to tell me the complete types of any one (or more) input files I ask for, without requiring stricter checks across all of their dependencies.  But also this isn't too hard to workaround, so I can totally understand not assigning this high priority.

**Comment by shiju001** [comment]
11
